### PR TITLE
Charlesmchen/rework system messages

### PR DESCRIPTION
**Pull request by charlesmchen**
This PR is now ready for review.

PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
This PR is now ready for review.

PTAL @michaelkirk

**Comment by charlesmchen** [comment]
screen shot 2017-06-05 at 4 12 33 pm
screen shot 2017-06-05 at 3 52 19 pm
screen shot 2017-06-05 at 3 04 55 pm
screen shot 2017-06-05 at 3 04 51 pm
screen shot 2017-06-05 at 3 04 38 pm

File: Signal/Images.xcassets/empty-group-avatar.imageset/Contents.json [hunk 1]
```diff
@@ -2,7 +2,16 @@
...
+    {
+      "idiom" : "universal",
+      "scale" : "2x"
+    },
+    {
+      "idiom" : "universal",
+      "scale" : "3x"
     }
   ],
   "info" : {
```
Comment (charlesmchen) [inline Signal/Images.xcassets/empty-group-avatar.imageset/Contents.json]: We could introduce separate flavors of empty-group-avatar for the various contexts in which it is used.

File: Signal/src/Models/OWSMessagesBubblesSizeCalculator.m [hunk 2]
```diff
@@ -76,6 +89,46 @@ - (CGSize)messageBubbleSizeForMessageData:(id<JSQMessageData>)messageData
...
+    OWSAssert(interaction);
+    OWSAssert(cacheKey);
+
+    NSValue *cachedSize = [self.cache objectForKey:cacheKey];
+    if (cachedSize != nil) {
+        return [cachedSize CGSizeValue];
+    }
+
+    CGSize result =
+        [OWSSystemMessageCell cellSizeForInteraction:interaction collectionViewWidth:layout.collectionView.width];
```
Comment (charlesmchen) [inline Signal/src/Models/OWSMessagesBubblesSizeCalculator.m]: The sizing logic has been moved into the cells for the systems messages and the unread indicator, to gather concerns and to leverage the constants defined there.

File: Signal/src/ViewControllers/DebugUI/DebugUIMessages.m [hunk 3]
```diff
@@ -117,22 +122,60 @@ + (OWSTableSection *)sectionForThread:(TSThread *)thread
                                        actionBlock:^{
                                            [DebugUIMessages sendMediaAttachments:100 thread:thread];
                                        }],
+                       [OWSTableItem itemWithTitle:@"Create all system messages"
+                                       actionBlock:^{
+                                           [DebugUIMessages createSystemMessagesInThread:thread];
+                                       }],
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/DebugUI/DebugUIMessages.m]: I added debug UI to create system messages.

File: Signal/src/ViewControllers/ConversationView/MessagesViewController.m [hunk 4]
```diff
@@ -3908,8 +3827,11 @@ - (void)popAllConversationSettingsViews
 
 - (BOOL)isSpecialItemAtIndexPath:(NSIndexPath *)indexPath
 {
+    // TODO:
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/MessagesViewController.m]: TODO what?

File: Signal/src/ViewControllers/ConversationView/MessagesViewController.m [hunk 5]
```diff
@@ -3908,8 +3827,11 @@ - (void)popAllConversationSettingsViews
 
 - (BOOL)isSpecialItemAtIndexPath:(NSIndexPath *)indexPath
 {
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/ConversationView/MessagesViewController.m]: so is "special" anything which is not a text message or attachment?

File: Signal/src/ViewControllers/DebugUI/DebugUIMessages.m (hunk 3 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/DebugUI/DebugUIMessages.m]: 👏 👏 👏

File: Signal/src/ViewControllers/DebugUI/DebugUIMessages.m [hunk 6]
```diff
@@ -117,22 +122,60 @@ + (OWSTableSection *)sectionForThread:(TSThread *)thread
...
         return;
     }
-    [ThreadUtil
-        sendMessageWithText:[[@(counter) description]
-                                stringByAppendingString:@" Lorem ipsum dolor sit amet, consectetur adipiscing elit. "
-                                                        @"Suspendisse rutrum, nulla vitae pretium hendrerit, tellus "
-                                                        @"turpis pharetra libero, vitae sodales tortor ante vel sem."]
-                   inThread:thread
-              messageSender:messageSender];
+    [self sendTextMessageInThread:thread counter:counter];
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/DebugUI/DebugUIMessages.m]: nit: These methods could benefit from some more distinct names.

`sendTextMessage:(int)counter thread:(TSThread *)thread`
`sendTextMessageInThread:(TSThread *)thread counter:(int)counter`

File: Signal/src/ViewControllers/DebugUI/DebugUITableViewController.m [hunk 7]
```diff
@@ -28,15 +28,38 @@ - (NSString *)tag
 
 #pragma mark - Factory Methods
 
+- (void)pushPageWithSection:(OWSTableSection *)section
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/DebugUI/DebugUITableViewController.m]: 👍

File: Signal/src/ViewControllers/OWSConversationSettingsTableViewController.m [hunk 8]
```diff
@@ -527,6 +485,31 @@ - (void)updateTableContents
     [self.tableView reloadData];
 }
 
+- (UITableViewCell *)disclosureCellWithName:(NSString *)name iconName:(NSString *)iconName
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/OWSConversationSettingsTableViewController.m]: nice cleanup.

File: Signal/src/views/OWSSystemMessageCell.m [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        [self.contentView addSubview:self.imageView];
+
+        self.titleLabel = [UILabel new];
+        self.titleLabel.textColor = [UIColor colorWithRGBHex:0x403e3b];
+        self.titleLabel.font = [OWSSystemMessageCell titleFont];
+        self.titleLabel.numberOfLines = 0;
+        self.titleLabel.lineBreakMode = NSLineBreakByWordWrapping;
+        [self.contentView addSubview:self.titleLabel];
+    }
+
```
Comment (michaelkirk) [inline Signal/src/views/OWSSystemMessageCell.m]: nit: Do we want to `OWSAssert(self.interaction?)` or we could make this method `configureWithInteraction:`

File: Signal/src/views/OWSSystemMessageCell.m [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+        OWSFail(@"Unknown interaction type");
+        return nil;
+    }
+    OWSAssert(result);
+    return result;
+}
+
++ (NSString *)titleForInteraction:(TSInteraction *)interaction
+{
+    // TODO: Should we move the copy generation into this view?
```
Comment (michaelkirk) [inline Signal/src/views/OWSSystemMessageCell.m]: Seems like a good idea to move it out of SSK models and towards the view, but could be another PR.

File: Signal/src/views/OWSSystemMessageCell.m [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+- (UIImage *)iconForInteraction:(TSInteraction *)interaction
+{
+    UIImage *result = nil;
+
+    if ([interaction isKindOfClass:[TSErrorMessage class]]) {
+        switch (((TSErrorMessage *)self.interaction).errorType) {
+            case TSErrorMessageInvalidKeyException:
+            case TSErrorMessageNonBlockingIdentityChange:
+            case TSErrorMessageWrongTrustedIdentityKey:
+            case TSErrorMessageMissingKeyId:
```
Comment (michaelkirk) [inline Signal/src/views/OWSSystemMessageCell.m]: Converging with the Android design language, I think we should have this shield icon used *only* for the identity changes. 

The other two (TSErrorMessageInvalidKeyException, TSErrorMessageMissingKeyId) indicate an error in client/service interaction, so I think they should be grouped with the `system_message_info`

File: Signal/src/views/OWSSystemMessageCell.m [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+            case TSErrorMessageInvalidMessage:
+            case TSErrorMessageDuplicateMessage:
+            case TSErrorMessageInvalidVersion:
+            case TSErrorMessageUnknownContactBlockOffer:
+                result = [UIImage imageNamed:@"system_message_info"];
+                break;
+        }
+    } else if ([interaction isKindOfClass:[TSInfoMessage class]]) {
+        switch (((TSInfoMessage *)self.interaction).messageType) {
+            case TSInfoMessageUserNotRegistered:
```
Comment (michaelkirk) [inline Signal/src/views/OWSSystemMessageCell.m]: nit: any reason to have this separate from the `system_message_info` in the next block?

File: Signal/src/ViewControllers/ConversationView/MessagesViewController.m (hunk 5 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/MessagesViewController.m]: I'm renaming this to `isUserMessageAtIndexPath` and adding a comment.

File: Signal/src/ViewControllers/ConversationView/MessagesViewController.m (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/MessagesViewController.m]: Fixed.

File: Signal/src/ViewControllers/DebugUI/DebugUIMessages.m (hunk 6 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/DebugUI/DebugUIMessages.m]: Renamed `sendTextMessage` to `sendTextMessages` (plural).

File: Signal/src/ViewControllers/DebugUI/DebugUITableViewController.m (hunk 7 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/DebugUI/DebugUITableViewController.m]: I'm eventually going to do more in this vein.

File: Signal/src/views/OWSSystemMessageCell.m (hunk 9 above)
Comment (charlesmchen) [inline Signal/src/views/OWSSystemMessageCell.m]: Done.

File: Signal/src/views/OWSSystemMessageCell.m (hunk 11 above)
Comment (charlesmchen) [inline Signal/src/views/OWSSystemMessageCell.m]: Done.

File: Signal/src/views/OWSSystemMessageCell.m (hunk 12 above)
Comment (charlesmchen) [inline Signal/src/views/OWSSystemMessageCell.m]: Combined.

File: Signal/src/views/OWSSystemMessageCell.m (hunk 10 above)
Comment (charlesmchen) [inline Signal/src/views/OWSSystemMessageCell.m]: I'd like to, but I suspect that a bunch of our logging logic is using that method. And it'll be non-trivial to find all of the cases where that usage is being done since they overloaded `[NSObject description]`.  I avoid overloading `[NSObject description]`.

Comment (michaelkirk) [inline Signal/src/views/OWSSystemMessageCell.m]: Agreed.
